janusgraph-core/src/main/java/org/janusgraph/graphdb/database/idhandling/IDHandler.java

```diff
@@ -38,6 +39,15 @@ public class IDHandler {
     public static final StaticBuffer MIN_KEY = BufferUtil.getLongBuffer(0);
     public static final StaticBuffer MAX_KEY = BufferUtil.getLongBuffer(-1);
 
+    public static final byte STOP_MASK = -128;
+
+    /**
+     * See {@link VariableString} for more detail. This marker has highest
+     * bit being one. When JanusGraph sees it, it knows the buffer stores
+     * a string ID rather than a positive long ID (whose highest bit is zero).
+     */
+    public static final byte STRING_ID_MARKER = -128;
```

Comment:
> then it seems it doesn't matter what this byte has

It does matter. I understand your confusion. Unfortunately, yes, this is indeed confusing. You might want to read the javadoc of `IDManager::getKeyID` to learn more about the background.

Briefly speaking, due to historical reasons, when a long-type vertex id appears as a partition key (`key` column in the data layout), it's always 8 bytes. It does use some weird encoding (which seems unnecessary to me) so that the first bit can be either 0 or 1 even though the value is always positive. Anyways, it always uses an 8-byte buffer to store the vertex ID (we could have just stored the value as it is without worrying about encoding and decoding - I don't know why this decision was made). In this case, the only way to distinguish a long-type ID and a non-long-type ID is to count the buffer size. Thus, we rely on the fact that 8 bytes are always `Long`. If we only support string type, then we don't need a marker at all. However, since we want to leave some space for future extensions, we use an arbitrary byte here. You are right the first bit can be anything.

When a long-type vertex id appears in the `column` column or `value` column, it no longer has a fixed length. It uses another encoding scheme that makes every but the last byte have the highest bit being 0. You might recall the usage of `b | STOP_MASK` or equivalently, `b | 0x80` that makes the last byte have a leading 1. Since long-type IDs are no longer 8 bytes, the above trick does not work anymore. Luckily, we leverage the fact that the highest bit of the first byte is always 0. Therefore, we use this `STRING_ID_MARKER` whose highest bit is 0. For now, this is just an arbitrary byte (0b1000000000000000). The last 15 bits are reserved for future usage.

To conclude, in the first case, the first byte can be anything. In the second case, the first byte can be anything except the highest bit MUST be 1. I use the same byte coz why not?

Back to your point:

> We probably could use lowest 4 bits to represent data type

I agree. That being said, we could defer this discussion until we actually start to support other types.

Reply:
Agree, but it could be a breaking change then. I thought that we could agree on the bits usage of this market and start using correct bits for that.
As you say the highest bit should always be 1 for “other type” and “0” for Long - so be it. We have plenty of bits in the market to be used.
We could do the next encoding for the marker:
highest bit - always 1.
next 2 bits - reserved for some latter features / extensions.
1 bit - tells if the original value is 7 bits or less. “1” - value is 7 bits or less. “0” - value is 8 bits or more.
4 bits - encoded data type. “0000” - is for String lets say.

Reply:
> At this point, I realize and regret that I should have written a design doc for better understanding at the beginning. Sorry for that.

No worries at all. I just ask some dummy questions because I didn’t work with that part of the codebase and simply don’t understand some parts sometimes. It’s great explanation though 👍 Thank you